docs/proposal/20211207-delegated-storage.md

```diff
@@ -0,0 +1,554 @@
+
+    ```yaml
+    id: a36ed738-dfe7-45ec-acd1-8e44e8db893b
+    name: capact-redis
+    typeRef:
+        path: cap.core.type.hub.storage.redis
+        revision: 0.1.0
+    latestResourceVersion:
+      metadata:
+        attributes:
+        - path: cap.core.attribute.default # if more such Typeinstances with default Attribute, select first one
+          revision: 0.1.0 
+        - path: cap.core.attribute.hub.storage.backend
+          revision: 0.1.0
+      value:
+        url: "storagebackend-handlers.capact-system:50051"
+        additionalParameters: null
+    backend: builtin # the storage option which stores already all other metadata. It could be the same Redis instance
+    ```
+
+    Default storage backend should have `additionalParameters` empty, or, optional.
+
+    **TODO:** How to enforce that default backend doesn't have any additionalParameters needed? Maybe we shouldn't enforce that at all, as there could be additionalParameters, but not required. We could enforce that if it is implemented on Policy level...
+
+## Workflow syntax - Create
+
+1. In workflow, Content Developer can specify requirements for a given backend:
+
+    ```yaml
+    requires:
+      cap.core.type.hub.storage: # Content Dev specifies such requirement only if he/she wants to force use a given backend
+        allOf:
+          - typeRef:
+              path: cap.type.helm.storage
+              revision: 0.1.0
+              alias: helm-storage
+    ```
+
+    - This workflow cannot be run unless there is a `helm-release` StorageBackend installed (where `helm-release` is only workflow alias).
+    - If there are no specific storage backend requirements set, the default backend will be used.
+
+1. Content Developer outputs the following Argo workflow artifact:
+
+    > **NOTE:** Before this proposal, the whole Argo workflow artifact was treated as a value. Now we would need to change that.
+
+    As value (if he/she uses default backend or backend without any required additional params):
+
+    ```yaml
+    value: foo # option 1: save a specific value on an external backend
+    ```
+
+    or
+
+    ```yaml
+    additionalParameters: # option 2: register something which already exist as external TypeInstance - based on additionalParameters
+      # however, additionalParameters are backend-specific properties, which means Content Dev need to explicitly specify the backend as described later.
+      name: release-name
+      namespace: release-namespace
+    ```
+
+    or both:
+
+    ```yaml
+    value: foo # option 3: save a specific value on an external backend with some additional parameters
+    additionalParameters:
+      key: bar
+      value: baz
+    ```
+
+    In that way, someday we will be able to extend such approach with additional properties:
+    ```yaml    
+    instrumentation: # someday - if we want to unify the approach
+      health:
+        endpoint: foo.bar/healthz
+      # (...)
+    ```
```

Comment:
but where this will be stored then? 

If you have the `helm-storage` backend and also add the **instrumentation** property. The `instrumentation` will be "extracted" from submitted TI and stored in our backend? What if a user wants to put there some sensitive data and store it in e.g. vault?

but I'm curious about your opinion, how we can handle that with this syntax

Reply:
Nope, the same backend as the value. I put that in the proposal.

Such `instrumentation` data would be also stored in the same storage backend as the `value`. If Content Developer wants to store it somewhere else, then an additional Argo artifact to produce is needed.